c++/src/capnp/compat/json.c++

```diff
@@ -380,72 +380,135 @@ void JsonCodec::encodeField(StructSchema::Field field, DynamicValue::Reader inpu
+  KJ_REQUIRE(endPtr == s.end(), "String does not contain valid floating number", s);
+  return value;
+}
+
+template <typename SetFn, typename DecodeArrayFn, typename DecodeObjectFn>
+void decodeField(Type type, JsonValue::Reader value, SetFn setFn, DecodeArrayFn decodeArrayFn,
+    DecodeObjectFn decodeObjectFn) {
+  //This code relies on conversions in DynamicValue::Reader::as<T>.
+  switch(type.which()) {
+    case schema::Type::VOID:
+      break;
+    case schema::Type::BOOL:
+      if (value.isBoolean()) {
+        setFn(value.getBoolean());
+      } else {
+        KJ_FAIL_REQUIRE("Expected boolean value");
+      }
       break;
-    case JsonValue::BOOLEAN:
-      set(value.getBoolean());
+    case schema::Type::INT8:
+    case schema::Type::INT16:
+    case schema::Type::INT32:
+    case schema::Type::INT64:
+      //Rellies on range check in DynamicValue::Reader::as<IntType>
+      if (value.isNumber()) {
+        setFn(value.getNumber());
+      } else if (value.isString()) {
+        setFn(parseInt64(value.getString()));
+      } else {
+        KJ_FAIL_REQUIRE("Expected numeric value");
+      }
+      break;
+    case schema::Type::UINT8:
+    case schema::Type::UINT16:
+    case schema::Type::UINT32:
+    case schema::Type::UINT64:
+      //Rellies on range check in DynamicValue::Reader::as<IntType>
+      if (value.isNumber()) {
+        setFn(value.getNumber());
+      } else if (value.isString()) {
+        setFn(parseUInt64(value.getString()));
+      } else {
+        KJ_FAIL_REQUIRE("Expected numeric value");
+      }
       break;
-    case JsonValue::NUMBER:
-      set(value.getNumber());
+    case schema::Type::FLOAT32:
+    case schema::Type::FLOAT64:
+      if (value.isNull()) {
+        setFn(kj::nan());
+      } else if (value.isNumber()) {
+        setFn(value.getNumber());
+      } else if (value.isString()) {
+        setFn(parseFloat64(value.getString()));
+      } else {
+        KJ_FAIL_REQUIRE("Expected numeric value");
+      }
       break;
-    case JsonValue::STRING:
-      if (type.isInt8() || type.isInt16() || type.isInt32() || type.isInt64()) {
-        set(parseInt64(value.getString()));
-      } else if (type.isUInt8() || type.isUInt16() || type.isUInt32() || type.isUInt64()) {
-        set(parseUInt64(value.getString()));
+    case schema::Type::TEXT:
+      if (value.isString()) {
+        setFn(value.getString());
       } else {
-        set(value.getString());
+        KJ_FAIL_REQUIRE("Expected string value");
       }
       break;
-    case JsonValue::ARRAY: {
-      if (type.isData()) {
-        kj::Vector<byte> data;
-        for (auto arrayObject : value.getArray()) {
+    case schema::Type::DATA:
+      if (value.isArray()) {
+        auto array = value.getArray();
+        kj::Vector<byte> data(array.size());
+        for (auto arrayObject : array) {
           auto x = int(arrayObject.getNumber());
           KJ_REQUIRE(x >= 0 && x <= 255, "Number in array of bytes out of range.");
```

Comment:
That would additionally catch non-integral values. Error message would have to change a tiny bit.